src/layers/doubleSymLayer.jl

```diff
@@ -170,43 +170,46 @@ function JthetaTmv(this::DoubleSymLayer{T},Z::Array{T},dummy::Array{T},theta::Ar
     return dtheta
 end
 
-function JYTmv(this::DoubleSymLayer{T},Z::Array{T},dummy::Array{T},theta::Array{T},Y::Array{T},tmp)  where {T<:Number}
+function JYTmv(this::DoubleSymLayer{T},Zin::Array{T},dummy::Array{T},theta::Array{T},Y::Array{T},tmp)  where {T<:Number}
 
     nex       = div(length(Y),nFeatIn(this))
-    Z         = reshape(Z,:,nex)
+    Z         = reshape(Zin,:,nex)
     th1,th2,th3,th4  = splitWeights(this,theta)
     Kop       = getOp(this.K,th1)
     A,dA      = this.activation(tmp[2],true)
 
     dAZ       = dA.*(Kop*Z)
     dAZ       = JYTmv(this.nLayer,dAZ,(T)[],th4,Kop*Y,tmp[1])
-    dAZ       = reshape(dAZ,:,nex)
-    dY  = -(Kop'*dAZ)
+    dAZ_out       = reshape(dAZ,:,nex)
+    dY  = -(Kop'*dAZ_out)
     return dY
 end
 
-function JTmv(this::DoubleSymLayer{T},Z::Array{T},dummy::Array{T},theta::Array{T},Y::Array{T},tmp)  where {T<:Number}
+function JTmv(this::DoubleSymLayer{T}, Zin::Array{T}, dummy::Array{T},
+                                            theta::Array{T}, Yin::Array{T}, tmp) where {T<:Number}
 
-    dY = (T)[]
-    nex       = div(length(Y),nFeatIn(this))
-    Z         = reshape(Z,:,nex)
-    Yt        = reshape(tmp[2],:,nex)
-    Y         = reshape(Y,:,nex)
-    th1, th2,th3,th4  = splitWeights(this,theta)
-    Kop       = getOp(this.K,th1)
-    A,dA    = this.activation(Yt,true)
+    nex  = div(length(Yin),nFeatIn(this))
+    Z         = reshape(Zin, :, nex)
+    Yt        = reshape(tmp[2]::Array{T,2},:,nex)
+    Y         = reshape(Yin,:,nex)
+    th1, th2, th3, th4  = splitWeights(this,theta)
+    #Kop       = getOp(this.K,th1)
+    A::Array{T,2}, dA::Array{T,2}    = this.activation(Yt,true)
 
     dth3      = vec(sum(this.Bout'*Z,2))
-    dAZ       = dA.*(Kop*Z)
-    dth2      = vec(sum(this.Bin'*dAZ,2))
-    dth4,dAZ  = JTmv(this.nLayer,dAZ,zeros(T,0),th4,Kop*Y,tmp[1])
 
-    dth1      = JthetaTmv(this.K,dAZ,zeros(T,0),Y)
+    KopZ = Amv(this.K, th1, Z)
```

Comment:
@eldadHaber @lruthotto How attached are you two to the idea of have `getOp` return a `LinearOperator`? We find that the `LinearOperator` does not return a type stable result, even though the `prod` function it calls is stable.

I'm not sure what is confusing the compiler, but just skipping the `LinearOperator` and calling the function as done here on line 201 fixed the issue. Do you have any problem with us doing this?

Reply:
I'm not too attached to it although it makes the code more readable. If this is a problem with the LinearOperator package and you can reproduce this easily it might be worth reporting it there.